### PR TITLE
Remove SAO when mapgen limits are reached

### PR DESCRIPTION
**Pull request by nerzhul**
Addresses #5824

Look at https://youtu.be/bqAj0N0e-xk

### CODE REVIEW

**Comment by paramat** [comment]
You can't change 'blockpos over mapgen limit' to work in nodes because it works in block co-ordinates and other code relies on this, 'blockpos' terminology is used in the engine for block co-ordinates.

**Comment by paramat** [comment]
How about using https://github.com/minetest/minetest/blob/master/src/mapblock.h#L677

**Comment by paramat** [comment]
You are using the setting 'map gen limit', instead, 'max map gen limit' (31000) should be used, see my comment in the issue.

**Comment by paramat** [comment]
So i guess https://github.com/minetest/minetest/blob/master/src/mapblock.h#L677 cannot be used.
If you change the entity pos t block co-ordinates then you can use un unchanged 'blockpos ver mapgen limit'.

**Comment by paramat** [comment]
Map gen limit for a world is now per-world so the chances of someone reducing it later are small, so the work i described is not high priority.

**Comment by nerzhul** [comment]
objectpos_over_limit returns 310000 and my position outside limit is 309299

File: src/map.cpp [hunk 1]
```diff
@@ -1406,7 +1412,7 @@ bool ServerMap::initBlockMake(v3s16 blockpos, BlockMakeData *data)
 
 	// Do nothing if not inside mapgen limits (+-1 because of neighbors)
 	if (blockpos_over_mapgen_limit(full_bpmin) ||
-			blockpos_over_mapgen_limit(full_bpmax))
+		blockpos_over_mapgen_limit(full_bpmax))
```
Comment (paramat) [inline src/map.cpp]: 2 indents needed here.

**Comment by paramat** [comment]
Discussion http://irc.minetest.net/minetest-dev/2017-05-27#i_4944488 to 16:59

**Comment by paramat** [comment]
> objectpos_over_limit returns 310000 and my position outside limit is 309299

Yes see IRC logs for why http://irc.minetest.net/minetest-dev/2017-05-27#i_4944356 terrain always stops long before 31000.

**Comment by bigfoot547** [comment]
What if player leaves map bounds?

File: src/content_sao.cpp [hunk 2]
```diff
@@ -402,15 +402,27 @@ void LuaEntitySAO::step(float dtime, bool send_recommended)
...
 		m_env->getScriptIface()->luaentity_Step(m_id, dtime);
 	}
 
-	if(send_recommended == false)
+	// Remove LuaEntity outside of mapgen limits
+	{
+		ServerMap *map = dynamic_cast<ServerMap *>(&m_env->getMap());
+		assert(map);
+		if (!m_pending_deactivation && map->sao_pos_over_mapgen_limit(m_base_position)) {
+			errorstream << "Remove SAO " << m_id << "(" << m_init_name
```
Comment (bigfoot547) [inline src/content_sao.cpp]: Could you use warningstream on release builds, and errorstream in debug builds?

**Comment by paramat** [comment]
-1 for the moment as the function needs improvement.

IRC:

Also needs checking at negative world edge as behaviour will be different, but yes doesn't need to be completely accurate
On which side of the terrain edge should the sao limit be? perhaps it's better to remove just inside than just outside? to avoid stuff falling down endlessly over the edge
Yes 'blockpos over mapgen limit' will be very inaccurate because terrain can stop up to 96 nodes before the mapgen limit. so keeping the removal inside the terrain edge means it could happen up to 96 nodes inside the terrain edge. however this could be considered 'weirdness at world's edge' :]

File: src/content_sao.cpp (hunk 2 above)
Comment (paramat) [inline src/content_sao.cpp]: This should be warningstream i feel, notifying of a delete is not critical enough to have in red letters in chat.

**Comment by paramat** [comment]
Writing out a full new function makes it simpler and faster:
```
bool ServerMap::sao_pos_over_mapgen_limit(const v3f &p)
{
	const float sao_limit_bs = MYMIN(
		getMapgenParams()->mapgen_limit, MAX_MAP_GENERATION_LIMIT) * BS - 940.0f;
	return p.X < -sao_limit_bs ||
		p.X >  sao_limit_bs ||
		p.Y < -sao_limit_bs ||
		p.Y >  sao_limit_bs ||
		p.Z < -sao_limit_bs ||
		p.Z >  sao_limit_bs;
}
```
Ths ensures the SAO is removed before the terrain edge, but has the disadvantage of removing a SAO a long way from an edge depending on the 'mapgen limit' set., this will not be popular with those who have created a tiny world or those who are dropping items near an edge, only to lose them.

Perhaps we will need to precisely calculate the terrain edge and use that.

**Comment by paramat** [comment]
Linking this here for future use, this defines the chunk minpos for a given blockpos:
emerge.cpp L284
```
v3s16 EmergeManager::getContainingChunk(v3s16 blockpos, s16 chunksize)
{
	s16 coff = -chunksize / 2;
	v3s16 chunk_offset(coff, coff, coff);

	return getContainerPos(blockpos - chunk_offset, chunksize)
		* chunksize + chunk_offset;
}
```
and is called from:
map.cpp L1394
```
bool ServerMap::initBlockMake(v3s16 blockpos, BlockMakeData *data)
{
	s16 csize = getMapgenParams()->chunksize;
	v3s16 bpmin = EmergeManager::getContainingChunk(blockpos, csize);
	v3s16 bpmax = bpmin + v3s16(1, 1, 1) * (csize - 1);

	bool enable_mapgen_debug_info = m_emerge->enable_mapgen_debug_info;
	EMERGE_DBG_OUT("initBlockMake(): " PP(bpmin) " - " PP(bpmax));

	v3s16 extra_borders(1, 1, 1);
	v3s16 full_bpmin = bpmin - extra_borders;
	v3s16 full_bpmax = bpmax + extra_borders;

	// Do nothing if not inside mapgen limits (+-1 because of neighbors)
	if (blockpos_over_mapgen_limit(full_bpmin) ||
			blockpos_over_mapgen_limit(full_bpmax))
		return false;
```
...
...

**Comment by nerzhul** [comment]
maybe you can ajust the magic 940.0f value ? my formula touch the world borders in positive values (not tested negative atm)

File: src/content_sao.cpp (hunk 2 above)
Comment (nerzhul) [inline src/content_sao.cpp]: yeah i will change this before merge, it was for testing purpose, i think infostream is sufficient here

**Comment by paramat** [comment]
I tested your version, for positive edges the SAO limit was 5 nodes inside the terrain edge, for negative edges it was 5 nodes outside the terrain edge. But that was at 31000, with smaller 'map gen limit' your maths looks like it would produce changing behaviour.

Doing this the simple way means an error of up to 94 nodes.
The reasons i think it should be exact are:
1. It should be possible to create and play a 1-mapchunk world, or play normally very close to the edge, without losing a (potentially valuable or unique) item when dropping it..
2. If the SAO limit is outside the terrain edge then dropped items can fall endlessly beyond the edge, or fall into the void below a world, potentially causing problems.
3. The positive and negative limits can be calculated once as 'consts' so there is no performance loss compared to simpler code.

**Comment by nerzhul** [comment]
the constness doesn't improve in any meaning the performance, calcul is done everytime, only macro & hardcoded values are calculated at compilation time.

We should have the exact limit value, well calculated than my limit (which is more precise than yours)

**Comment by paramat** [comment]
Well i mean, a const value is calculated only once per game session instead of every server step, so it does not affect in-game performance.

**Comment by nerzhul** [comment]
This is totally wrong. Const is for constness, making a variable read only . Static is calculated one time per binary life and thread_local (C++11) 1 time per thread life

**Comment by paramat** [comment]
What i mean is, my understanding is that a const variable is calculated once the first time the line is run, then not calculated after that. Const variables are present in many performance critical locations and performance would suffer unnecessarily if the calculation was run every time the line is run.

**Comment by paramat** [comment]
Working in this now, very complicated, made worse by the fact that `blockpos_over_mapgen_limit()' is innaccurate, i made it simple when i worked on it which is making this PR work more complex.
I'm sure i can do this precisely but not sure it will be ready for release.

**Comment by paramat** [comment]
> Const is for constness, making a variable read only . Static is calculated one time per binary life and thread_local (C++11) 1 time per thread life

Should the following be static then? considering that it must be recalculated when exiting to menu and entering a new world.
```
bool ServerMap::blockpos_over_mapgen_limit(v3s16 p)
{
	const s16 mapgen_limit_bp = rangelim(
		getMapgenParams()->mapgen_limit, 0, MAX_MAP_GENERATION_LIMIT) /
		MAP_BLOCKSIZE;
	return p.X < -mapgen_limit_bp ||
		p.X >  mapgen_limit_bp ||
		p.Y < -mapgen_limit_bp ||
		p.Y >  mapgen_limit_bp ||
		p.Z < -mapgen_limit_bp ||
		p.Z >  mapgen_limit_bp;
}
```

**Comment by paramat** [comment]
I have positive limit precisely calculated, now for negative.

**Comment by paramat** [comment]
Got both, copying here for safety, this is rough testing code, it is so complex i think this will have to be in the constructor and the values made into class members.
```bool ServerMap::sao_pos_over_mapgen_limit(const v3f &p)
{
	s16 csize = getMapgenParams()->chunksize;  // in blocks
	s16 coff = -csize / 2;  // in blocks
	// minp in nodes of central chunk
	s16 ccminn = coff * MAP_BLOCKSIZE;
	// maxp in nodes of central chunk
	s16 ccmaxn = (coff + csize) * MAP_BLOCKSIZE - 1;
	// fullmaxp in nodes of central chunk
	s16 ccfmaxn = (coff + csize + 1) * MAP_BLOCKSIZE - 1;
	// fullminp in nodes of central chunk
	s16 ccfminn = (coff - 1) * MAP_BLOCKSIZE;
	// furthest block allowed, in blocks
	s16 mapgen_limit_bp = rangelim(
		getMapgenParams()->mapgen_limit, 0, MAX_MAP_GENERATION_LIMIT) /
		MAP_BLOCKSIZE;
	// furthest pos node allowed, in nodes
	s16 mapgen_limit_pos_n = (mapgen_limit_bp + 1) * MAP_BLOCKSIZE - 1;
	// furthest neg node allowed, in nodes
	s16 mapgen_limit_neg_n = -mapgen_limit_bp * MAP_BLOCKSIZE;
	// number of full chunks to furthest generated
	s16 nchunkspos = (mapgen_limit_pos_n - ccfmaxn) / (csize * MAP_BLOCKSIZE);
	// number of full chunks to furthest generated
	s16 nchunksneg = (ccfminn - mapgen_limit_neg_n) / (csize * MAP_BLOCKSIZE);
	// positive terrain edge in nodes
	s16 pos_edgen = ccmaxn + nchunkspos * csize * MAP_BLOCKSIZE;
	// negitive terrain edge in nodes
	s16 neg_edgen = ccminn - nchunksneg * csize * MAP_BLOCKSIZE;

	printf("pos edge: %d\n", pos_edgen);
	printf("neg edge: %d\n", neg_edgen);

	float sao_limit_bs_pos = pos_edgen * BS;
	float sao_limit_bs_neg = neg_edgen * BS;

	return p.X < sao_limit_bs_neg ||
		p.X > sao_limit_bs_pos ||
		p.Y < sao_limit_bs_neg ||
		p.Y > sao_limit_bs_pos ||
		p.Z < sao_limit_bs_neg ||
		p.Z > sao_limit_bs_pos;
}
```
I'll make a new PR.

**Comment by nerzhul** [comment]
no need for a new PR, just edit and commit on this PR i'm the fixer :)

**Comment by paramat** [comment]
Ok.

**Comment by paramat** [comment]
Sorry i don't know yet how to edit and take over someone elses PR.

**Comment by nerzhul** [comment]
you can either edit using github, or clone my repository with the pr branch and push to it:

git clone git@github.com:nerzhul/minetest.git -b sao_mapgen_limits

**Comment by paramat** [comment]
More testing, seems good.
At world base a dropped item will disappear the moment it passes into the void, nice.

**Comment by paramat** [comment]
Modified `sao_limt_bs_pos/neg` by 5.0f to make the limit absolutely exact to 0.1 nodes.

**Comment by paramat** [comment]
I have created a branch of my own for experiments, still trying to move that code into the serverMap constructor.

**Comment by nerzhul** [comment]
@paramat if you want some values store at runtime for mapgen limits, it's better to make it owned by Mapgen params or mapgen

**Comment by paramat** [comment]
Ok thanks.

**Comment by paramat** [comment]
Yes something as you suggest may be needed. When i try to calculate SAO limits in the serverMap constructor i get a segfault, the calculation probably relies on something done after this constructor.
So the best i can so (hacky) is here https://github.com/minetest/minetest/compare/master...paramat:worldedge2 i calculate SAO limits once only from `saoPosOverMapgenLimit(const v3f &p)`

**Comment by nerzhul** [comment]
It's very very hacky to test the 0.0f, instead use a bool m_sao_map_limit_generated and don't forget m_ for variables owned by the call. And do if (m_sao_map_limit_generated) { return ; } in the calculation function

**Comment by nerzhul** [comment]
i also prefer if this function is owned by mapgen params, as the limit is a mapgen param

**Comment by paramat** [comment]
> i also prefer if this function is owned by mapgen params, as the limit is a mapgen param

Yes. i will do this, i know what to do now.

**Comment by paramat** [comment]
New PR #5856

**Pull request closed without merging**